nameserver/mdns_server.go

```diff
@@ -50,23 +50,52 @@ func (s *MDNSServer) Start(ifi *net.Interface) error {
 		return err
 	}
 
-	handleMDNS := func(w dns.ResponseWriter, r *dns.Msg) {
-		// Ignore answers to other questions
-		if len(r.Answer) == 0 && len(r.Question) > 0 {
-			q := r.Question[0]
-			if ip, err := s.zone.LookupLocal(q.Name); err == nil {
-				m := makeAddressReply(r, &q, []net.IP{ip})
-				if err = s.sendResponse(m); err != nil {
-					Warning.Printf("Error writing to %s", w)
+	handler := func(qtype uint16, lookup func(*dns.Msg, *dns.Question) *dns.Msg) dns.HandlerFunc {
+		return func(w dns.ResponseWriter, r *dns.Msg) {
+			// Handle only questions, ignore answers. We might also
+			// ignore questions that arise locally (i.e., that come
+			// from an IP we think is local), but in the interest of
+			// avoiding complication, and easier testing, this is
+			// elided on the assumption that the client wouldn't ask
+			// if it already knew the answer, and if it does ask,
+			// it'll be happy to get an answer.
+			if len(r.Answer) == 0 && len(r.Question) > 0 {
+				q := &r.Question[0]
+				if q.Qtype == qtype {
+					if m := lookup(r, q); m != nil {
+						Debug.Printf("Found local answer to mDNS query %s", q.Name)
+						if err = s.sendResponse(m); err != nil {
+							Warning.Printf("Error writing to %s", w)
+						}
+					} else {
+						Debug.Printf("No local answer for mDNS query %s", q.Name)
+					}
 				}
-			} else if s.addrIsLocal(w.RemoteAddr()) {
-				// ignore this - it's our own query received via multicast
-			} else {
-				Debug.Printf("Failed MDNS lookup for %s", q.Name)
 			}
 		}
 	}
-	go dns.ActivateAndServe(nil, conn, dns.HandlerFunc(handleMDNS))
+
+	handleA := handler(dns.TypeA, func(r *dns.Msg, q *dns.Question) *dns.Msg {
+		if ip, err := s.zone.LookupLocal(q.Name); err == nil {
+			return makeAddressReply(r, q, []net.IP{ip})
+		} else {
+			return nil
+		}
+	})
+
+	handlePTR := handler(dns.TypePTR, func(r *dns.Msg, q *dns.Question) *dns.Msg {
```

Comment:
I think this should be named more about  reverse lookups and less about PTR.  There could be other PTR requests in future